### PR TITLE
MapSqlParameterSource.getValue Handling of null values

### PR DESCRIPTION
**Pull request by rengy-github**
Does the reading of multithreaded HashMap need to be synchronized? This problem has been bothering me. I don't think it's necessary

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@rengy-github Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=24524)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-framework&pullRequestId=24524) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@rengy-github Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=24524)!

**Comment by rengy-github** [comment]
Using synchronized only with put improves performance

File: spring-jdbc/src/main/java/org/springframework/jdbc/core/namedparam/NamedParameterJdbcTemplate.java [hunk 1]
```diff
@@ -429,9 +429,17 @@ protected ParsedSql getParsedSql(String sql) {
 		if (getCacheLimit() <= 0) {
 			return NamedParameterUtils.parseSqlStatement(sql);
 		}
-		synchronized (this.parsedSqlCache) {
-			return this.parsedSqlCache.computeIfAbsent(sql, NamedParameterUtils::parseSqlStatement);
+		ParsedSql parsedSql = this.parsedSqlCache.get(sql);
+		if (parsedSql == null) {
+			synchronized (this.parsedSqlCache) {
+				parsedSql = this.parsedSqlCache.get(sql);
```
Comment (hyeonisism) [inline spring-jdbc/src/main/java/org/springframework/jdbc/core/namedparam/NamedParameterJdbcTemplate.java]: I think you can change the code more clearly with the computeIfAbsent() method.

I'd recommend you refer to [this](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-).

**Comment by rengy-github** [comment]
I know it's not thread safe to use synchronization only when putting，But occasionally SQL is parsed twice without any problem

**Comment by rengy-github** [comment]
i think parse a sql multiple times   better than blocking   other thread while get sql

**Comment by rengy-github** [comment]
My puzzle is whether there is problem with the other thread get when one thread puts

Could send me a link about this, thank you

**Comment by snicoll** [comment]
There has been a number of changes that leave the current proposal of returning null if the key is not set, which is a breaking change.

**Pull request closed without merging**